Add migration guide from ceph-ansible to cephadm (#957)
osism/osism.github.io: docs/guides/migration-guide/cephadm/index.md

---
sidebar_label: Cephadm
sidebar_position: 32
---

# Migrate from Ceph-Ansible to Cephadm

:::warning

The migration from ceph-ansible to cephadm is performed _in place_ by adopting existing

> Ceph daemons one at a time. While this process is designed to be non-disruptive, it is

Reviewer: It may be puzzling to some users why there is https://osism.tech/docs/guides/migration-guide/rook/ "to migrate from Ceph-Ansible to Rook" _and_ https://osism.tech/docs/guides/migration-guide/cephadm/  for the "migration from ceph-ansible to cephadm".

Author: Rook migration guide will be removed.

**strongly advised** to test the migration in a controlled environment first, such as the
[OSISM testbed](https://github.com/osism/testbed). Ensure that precautionary backups are
made and all other necessary safety measures are in place before migrating a production
cluster.

TODO: Add specific guidance on recommended backup strategies (e.g. cluster-wide backups
to a separate cluster) and concrete safety measures to take before starting the migration.

> 

Reviewer: The TODO at lines 17–18 is covered by the Known limitations section: "Backup and safety measures: Specific guidance on recommended backup strategies and concrete pre-migration safety measures is still being prepared." Suggest removing the duplicate:

```suggestion
```

* **Automated readiness checks**: A planned `osism apply ready-for-cephadm` task to automate prerequisite verification is not yet available.

This note will be updated as additional sections are completed.

:::

## Background

The deployment tool ceph-ansible is deprecated as of OSISM 10 and will not be supported
in upcoming OSISM releases. The
[official recommendation](https://github.com/ceph/ceph-ansible/blob/main/README.rst) from
the Ceph project is to migrate to [cephadm](https://docs.ceph.com/en/latest/cephadm/).

After migration, Ceph daemons run as containers managed by cephadm instead of
ceph-ansible. All lifecycle operations (upgrades, expansions, configuration
changes) are then performed through cephadm and the Ceph orchestrator.

For the full upstream documentation, refer to
[Switching from ceph-ansible to cephadm](https://docs.ceph.com/en/latest/cephadm/adoption/).

## Prerequisites

* A running Ceph cluster deployed with ceph-ansible via OSISM.
* All Ceph daemons are healthy (`ceph -s` reports `HEALTH_OK` or only expected warnings).

> * SSH access to all Ceph nodes from the node where cephadm will be run.

Reviewer: ```suggestion
* SSH access from the OSISM manager node to all Ceph nodes (required by cephadm for orchestration).
```

* The Ceph cluster must be running at least Ceph Octopus (15.2.x) or later. Clusters on
  OSISM 7 or later already meet this requirement.
* Python 3 and `lvm2` must be installed on all Ceph nodes (these are typically already present).

TODO: Consider replacing the manual prerequisite checks with an `osism apply ready-for-cephadm`
task that automatically verifies all conditions (cluster health, SSH access, Ceph version,
required packages).

## Step 1: Verify cluster health

Before starting the migration, ensure the cluster is in a healthy state. Run the following

> commands on the OSISM manager node .

Reviewer: ```suggestion
commands on the OSISM manager node.
```

```bash
ceph -s
ceph osd tree
ceph df
```

> All PGs should be `active+clean`. Resolve any degraded or misplaced PGs before proceeding.

Reviewer: ```suggestion
All PGs should be `active+clean`. Resolve any degraded or misplaced PGs before proceeding.
In `ceph osd tree`, verify that all OSDs show `up` with a non-zero REWEIGHT. In `ceph df`,
check that `%RAW USED` is well below 85% (Ceph's default nearfull threshold).
```

TODO: Consider including the health verification commands in the `osism apply ready-for-cephadm`
task as well.

> 

Reviewer: The wording "consider including ... as well" implies the command already exists but is incomplete — but `osism apply ready-for-cephadm` does not exist at all. Suggest removing this TODO:

```suggestion
```

## Step 2: Install cephadm

Install cephadm on **all** Ceph nodes. The version of cephadm must match the running Ceph
release.

First, determine the running Ceph release on the OSISM manager node :

```bash
ceph version

> ```
> 
> This returns output like `ceph version 18.2.7 (...)  reef (stable)`. The full version
> number (e.g. `18.2.7`) is needed for the next step.
> 
> ### Control nodes (mon, mgr)
> 
> Install the cephadm package from the

Reviewer: ```suggestion
### Control nodes (mon, mgr)

:::note
OSD nodes require a different installation method due to a known bug in the UCA cephadm
package for Reef. See the [OSD nodes](#osd-nodes) section below before installing on OSD nodes.
:::

Install the cephadm package from the
```

[Ubuntu Cloud Archive](https://ubuntu-cloud.archive.canonical.com/ubuntu/pool/main/c/ceph/)
on the control nodes:

| Ceph Release | cephadm Package                                    |
|:-------------|:---------------------------------------------------|
| Quincy (17)  | `cephadm_17.2.9-0ubuntu0.22.04.1~cloud0_amd64.deb` |
| Reef (18)    | `cephadm_18.2.4-0ubuntu1~cloud1_amd64.deb`         |
| Squid (19)   | `cephadm_19.2.3-0ubuntu0.24.04.2~cloud0_amd64.deb` |

These are the only versions available in the UCA. The exact point release does not need
to match your running Ceph version -- only the release series (Quincy, Reef, Squid) matters.

The package is installed directly via `dpkg` rather than `apt install` because the
Ubuntu Cloud Archive is typically not configured as an apt source in OSISM environments.
Adding the full UCA repository just for cephadm is not recommended, as it could
introduce unintended package upgrades from the UCA that conflict with the versions
managed by OSISM.

```bash
CEPHADM_PKG=cephadm_18.2.4-0ubuntu1~cloud1_amd64.deb
```

```bash
curl --silent --remote-name --location https://ubuntu-cloud.archive.canonical.com/ubuntu/pool/main/c/ceph/${CEPHADM_PKG}

> sudo dpkg -i ${CEPHADM_PKG}
> ```
> 
> ### OSD nodes
> 
> The UCA cephadm package for Reef (18.2.4) contains a
> [known bug](https://tracker.ceph.com/issues/63151) that causes a crash when parsing
> AppArmor profiles during OSD adoption. This was
> [fixed upstream](https://github.com/ceph/ceph/pull/57955) and is included in Reef
> v18.2.5+, but the UCA has not updated the packaged version beyond 18.2.4.
> 
> On OSD nodes, install cephadm as a standalone Python script from the Ceph Git
> repository instead to get a version that includes the fix:

Reviewer: ```suggestion
### OSD nodes

If you are running **Quincy or Squid**, install cephadm on OSD nodes using the same UCA
package as described for control nodes above.

If you are running **Reef (18.x)**, the UCA cephadm package (18.2.4) contains a
[known bug](https://tracker.ceph.com/issues/63151) that causes a crash when parsing
AppArmor profiles during OSD adoption. This was
[fixed upstream](https://github.com/ceph/ceph/pull/57955) and is included in Reef
v18.2.5+, but the UCA package has not been updated beyond 18.2.4.

On Reef OSD nodes, install cephadm as a standalone Python script from the Ceph Git
repository instead to get a version that includes the fix:
```

```bash
CEPH_RELEASE=$(docker inspect $(docker ps --filter "name=ceph" --format "{{.Names}}" | head -1) --format '{{.Config.Image}}' | cut -d: -f2)
curl --silent --remote-name --location https://raw.githubusercontent.com/ceph/ceph/${CEPH_RELEASE}/src/cephadm/cephadm.py

> chmod +x cephadm.py

Author: Add missing v here: https://github.com/ceph/ceph/tags

  Ceph nodes can cause monitors to lose quorum
* General system prerequisites such as the availability of `systemctl`

If any of these checks fail, `prepare-host` will report the issue so it can be resolved
before proceeding with the migration.

```bash
sudo cephadm prepare-host
```

The output should look similar to this:

```console
Verifying podman|docker is present...
Verifying lvm2 is present...
Verifying time synchronization is in place...
Unit chrony.service is enabled and running
Repeating the final host check...
docker (/usr/bin/docker) is present
systemctl is present
lvcreate is present
Unit chrony.service is enabled and running
Host looks OK
```

If the output indicates errors or missing dependencies, resolve them before proceeding.

Determine the currently used container image from a running Ceph daemon on
**one of the monitor nodes**:

```bash
CEPH_IMAGE=$(docker inspect $(docker ps --filter "name=ceph" --format "{{.Names}}" | head -1) --format '{{.Config.Image}}')
```

Set the container image in the Ceph configuration. Run the following command on the
OSISM manager node:

```bash

> ceph config set global container_image ${CEPH_IMAGE}

Reviewer: Running the `docker inspect` command on the manager node returns the `cephclient` image (`cephclient:reef`) instead of the correct Ceph daemon image (`ceph-daemon:reef`) — a plausible-looking wrong value. The variable is then unset when switching to the manager node to run `ceph config set`. Suggest doing everything from the manager via SSH:

```suggestion
Determine the currently used container image and set it in the Ceph configuration.
Run the following commands on the OSISM manager node:

```bash
MON_NODE=$(osism get hosts -l ceph-mon | awk 'NR>3 && /\|/ {print $2}' | head -1)
CEPH_IMAGE=$(ssh ${MON_NODE} "docker inspect \$(docker ps --filter 'name=ceph' --format '{{.Names}}' | head -1) --format '{{.Config.Image}}'")
echo "Container image: ${CEPH_IMAGE}"
ceph config set global container_image ${CEPH_IMAGE}
```

:::warning
Do not run the `docker inspect` command on the manager node — it runs the `cephclient`
container there and would return a wrong image (`cephclient:reef` instead of `ceph-daemon:reef`).
:::
```

```

Import the existing `ceph.conf` into the central monitor config store. This ensures
that custom tuning parameters are preserved after migration, as cephadm uses the
centralized config store instead of per-node `ceph.conf` files. Since the `ceph` CLI
is not installed on the host, execute the command inside the crash container on
**one of the monitor nodes**:

```bash
docker exec ceph-crash-$(hostname) ceph config assimilate-conf -i /etc/ceph/ceph.conf
```

In a typical OSISM deployment, the `ceph.conf` is identical across all nodes, so running
this once is sufficient. If nodes have individual tuning parameters in their `ceph.conf`,
run the command on each affected node.

## Step 4: Configure cephadm

Enable the cephadm orchestrator module on the OSISM manager node:

```bash
ceph mgr module enable cephadm
ceph orch set backend cephadm
```

Configure cephadm to use the `dragon` user (which has passwordless sudo) instead of
`root`. Import the existing operator SSH key so that cephadm can connect to all Ceph
nodes. On the OSISM manager node:

> 

Reviewer: ```suggestion
Configure cephadm to use the `dragon` user (which has passwordless sudo) instead of
`root`. In a standard OSISM deployment, the operator SSH key already exists at
`/opt/ansible/secrets/id_rsa.operator`. Import it so that cephadm can connect to all
Ceph nodes. On the OSISM manager node:
```

```bash
ceph cephadm set-user dragon
cp /opt/ansible/secrets/id_rsa.operator* /opt/cephclient/data/
ceph cephadm set-priv-key -i /data/id_rsa.operator
ceph cephadm set-pub-key -i /data/id_rsa.operator.pub
rm /opt/cephclient/data/id_rsa.operator*
```

> 

Author: Maybe add here how to generate id_rsa.operator.pub if only id_rsa.operator exists.

:::info

If no existing SSH key is available, generate a new one and distribute it to all Ceph
nodes instead:

```bash
ceph cephadm generate-key
ceph cephadm get-pub-key > /tmp/cephadm-pub-key.pub
```

Copy the public key to all Ceph nodes:

```bash
ssh-copy-id -f -i /tmp/cephadm-pub-key.pub dragon@<node>
```

Or use a loop to distribute the key to all Ceph nodes at once:

```bash
for node in $(osism get hosts -l ceph | awk 'NR>3 && /\|/ {print $2}'); do
    ssh-copy-id -f -i /tmp/cephadm-pub-key.pub dragon@${node}
done
```

:::

Register all Ceph nodes with the orchestrator on the OSISM manager node:

```bash
ceph orch host add <node> <node-ip>
```

Or use a loop to register all Ceph nodes at once:

```bash

> for node in $(osism get hosts -l ceph | awk 'NR>3 && /\|/ {print $2}'); do

Author: Check first if osism get hosts is usable (it's a pretty new command).

    ceph orch host add ${node} $(getent hosts ${node} | awk '{print $1}')
done
```

The output should look similar to this:

```console
Added host 'testbed-node-0' with addr '192.168.16.10'
Added host 'testbed-node-1' with addr '192.168.16.11'
Added host 'testbed-node-2' with addr '192.168.16.12'
Added host 'testbed-node-3' with addr '192.168.16.13'
Added host 'testbed-node-4' with addr '192.168.16.14'
Added host 'testbed-node-5' with addr '192.168.16.15'
```

Verify that all hosts have been registered:

```bash
ceph orch host ls
```

The output should look similar to this:

```console
HOST            ADDR           LABELS  STATUS
testbed-node-0  192.168.16.10
testbed-node-1  192.168.16.11
testbed-node-2  192.168.16.12
testbed-node-3  192.168.16.13
testbed-node-4  192.168.16.14
testbed-node-5  192.168.16.15
6 hosts in cluster
```

## Step 5: Adopt daemons

Adopt all Ceph daemons across the cluster. The recommended order is:

1. **Monitors** (mon)
2. **Managers** (mgr)
3. **OSDs** (osd)

> 
> For each daemon, run the adopt command on the respective node.
> 
> :::warning
> 
> Verify cluster health with `ceph -s` on the OSISM manager node after each step.
> Do not proceed if the cluster is not healthy.
> 
> :::

Reviewer: After registering hosts with cephadm, `ceph -s` will show `HEALTH_WARN` with "stray daemon(s) not managed by cephadm" — this is expected and does not mean the migration should stop. The current wording is too absolute:

```suggestion
:::warning

Verify cluster health with `ceph -s` on the OSISM manager node after each step.
Do not proceed if `ceph -s` reports `HEALTH_ERR` or degraded/unavailable PGs.

During migration, `HEALTH_WARN` with messages like "stray daemon(s) not managed by
cephadm" or "failed to probe daemons or devices" is expected and safe to continue.

:::
```

unit, migrates data and logs, and creates new cephadm-managed systemd units. Since
multiple managers run in active/standby mode, adopting one at a time ensures the
cluster always has an active manager available.

On each manager node, set the required variables and adopt the daemon:

```bash
CEPH_HOSTNAME=$(hostname)
CEPH_IMAGE=$(docker inspect $(docker ps --filter "name=ceph" --format "{{.Names}}" | head -1) --format '{{.Config.Image}}')
sudo cephadm --image ${CEPH_IMAGE} adopt --style legacy --skip-firewalld --name mgr.${CEPH_HOSTNAME}
sudo systemctl reset-failed ceph-mgr@${CEPH_HOSTNAME}.service 2>/dev/null || true
```

The output should look similar to this:

```console
Pulling container image registry.osism.tech/osism/ceph-daemon:reef...
Stopping old systemd unit ceph-mgr@testbed-node-0...
Disabling old systemd unit ceph-mgr@testbed-node-0...
Moving data...
Chowning content...
Moving logs...
Creating new units...
```

### Verify monitors and managers

After all monitors and managers have been adopted, verify on the OSISM manager node that
the orchestrator recognises them:

List all services known to the orchestrator:

```bash

> ceph orch ls

Reviewer: Without `--refresh` the count may show stale data (e.g. `2/0`) for several minutes after the last daemon is adopted:

```suggestion
ceph orch ls --refresh
```

real data volumes (BlueStore). The adopt process migrates each OSD's data directory,
block device symlinks, and logs into the cephadm layout — but the underlying data on
disk is **not** moved or modified.

Because an OSD restart temporarily reduces the number of available replicas, safety flags
(`noout`, `nodeep-scrub`, balancer off) are set beforehand to prevent Ceph from
initiating unnecessary data rebalancing or deep scrubs while OSDs are being restarted
during adoption. The PG autoscaler is also disabled to avoid placement group changes
during the process.

Before adopting OSDs, set safety flags on the OSISM manager node to prevent unnecessary data
movement and PG changes during the adoption process:

```bash
ceph osd set noout
ceph osd set nodeep-scrub
ceph balancer off
```

Disable the PG autoscaler on all pools that have it enabled. Record which pools had it
enabled so it can be re-enabled after adoption:

```bash
for pool in $(ceph osd pool ls); do
    mode=$(ceph osd pool get ${pool} pg_autoscale_mode -f json | python3 -c "import json,sys; print(json.load(sys.stdin)['pg_autoscale_mode'])")

>     if [ "${mode}" = "on" ]; then
>         echo "${pool}" >> /home/dragon/autoscale_pools.txt
>         ceph osd pool set ${pool} pg_autoscale_mode off
>     fi
> done
> ```
> 
> On each OSD node, set the required variable and identify the OSDs running on the node:
> 
> ```bash
> CEPH_IMAGE=$(docker inspect $(docker ps --filter "name=ceph" --format "{{.Names}}" | head -1) --format '{{.Config.Image}}')
> docker ps --filter "name=ceph-osd"
> ```
> 
> The output should look similar to this:
> 
> ```console
> CONTAINER ID   IMAGE                                       COMMAND                CREATED    STATUS    NAMES
> 05eb59c1ef36   registry.osism.tech/osism/ceph-daemon:reef  "/usr/bin/ceph-osd …"  5 days ago Up 5 days ceph-osd-3
> 8646edf83163   registry.osism.tech/osism/ceph-daemon:reef  "/usr/bin/ceph-osd …"  5 days ago Up 5 days ceph-osd-1
> ```
> 
> The OSD ID is the number after `ceph-osd-` in the container name. For example,
> `ceph-osd-1` has OSD ID `1` and `ceph-osd-3` has OSD ID `3`.
> 
> Then adopt each OSD on the node:

Reviewer: The warning about node-at-a-time adoption currently appears after both the per-OSD commands and the loop, so users may start adopting before reading it. Moving it here ensures they see the guidance before running anything. If this suggestion is accepted, remove the original warning block (see companion suggestion below).

```suggestion
`ceph-osd-1` has OSD ID `1` and `ceph-osd-3` has OSD ID `3`.

:::warning

Adopt OSDs **one node at a time**. After completing all OSDs on a node, wait until all
PGs return to `active+clean` (`ceph -s` shows all PGs `active+clean` under `data:`)
before proceeding to the next node. The overall health may still show `HEALTH_WARN` for
stray daemons during migration — this is expected and not a stop condition.

Within a single node, it is safe to use the loop below without waiting between individual
OSDs, provided your cluster meets both of the following conditions:

* At least **3 OSD nodes**.
* **Host-level CRUSH failure domain** — the OSISM default. Verify with:
  ```bash
  ceph osd crush rule dump
  ```
  All rules should show `"type": "host"` in the `chooseleaf` step.

If your cluster has only 2 OSD nodes, or any rule shows a different failure domain type,
adopt each OSD individually and wait for all PGs to return to `active+clean` after each
one before continuing.

:::

Then adopt each OSD on the node:
```

> 
> ```bash
> sudo cephadm --image ${CEPH_IMAGE} adopt --style legacy --skip-firewalld --name osd.<osd_id>
> sudo systemctl reset-failed ceph-osd@<osd_id>.service 2>/dev/null || true
> ```

Reviewer: ```suggestion
Then adopt each OSD on the node:

```bash
OSD_ID=<osd_id>
sudo cephadm --image ${CEPH_IMAGE} adopt --style legacy --skip-firewalld --name osd.${OSD_ID}
sudo systemctl reset-failed ceph-osd@${OSD_ID}.service 2>/dev/null || true
```
```

The output should look similar to this:

```console
Pulling container image registry.osism.tech/osism/ceph-daemon:reef...
Found online OSD at //var/lib/ceph/osd/ceph-1/fsid
objectstore_type is bluestore
Stopping old systemd unit ceph-osd@1...
Disabling old systemd unit ceph-osd@1...
Moving data...
Chowning content...
Chowning /var/lib/ceph/11111111-1111-1111-1111-111111111111/osd.1/block...
Disabling host unit ceph-volume@ lvm unit...
Non-zero exit code 1 from systemctl disable ceph-volume@lvm-1-f3cfe0e4-70f3-4078-9aba-2d45170e9df9.service
systemctl: stderr Failed to disable unit: Unit file ceph-volume@lvm-1-f3cfe0e4-70f3-4078-9aba-2d45170e9df9.service does not exist.
Moving logs...
Creating new units...
```

Or use a loop to adopt all OSDs on the current node at once:

```bash
for osd_id in $(docker ps --filter "name=ceph-osd" --format "{{.Names}}" | sed 's/ceph-osd-//'); do

>     sudo cephadm --image ${CEPH_IMAGE} adopt --style legacy --skip-firewalld --name osd.${osd_id}
>     sudo systemctl reset-failed ceph-osd@${osd_id}.service 2>/dev/null || true
> done
> ```
> 
> :::warning
> 
> Adopt OSDs **one node at a time**. After completing all OSDs on a node, wait until the
> cluster has fully recovered (`ceph -s` shows `HEALTH_OK`) before proceeding to the next
> node.
> 
> In **small environments** (e.g. 3 nodes with only 1–2 OSDs each), do **not** use the
> loop above. Instead, adopt each OSD individually and verify cluster health between each
> adoption. In small clusters, every single OSD represents a large fraction of the total
> data redundancy — briefly taking one offline during adoption may already bring the cluster
> to `min_size`. Adopting a second OSD before the first has fully rejoined could lead to
> degraded or unavailable placement groups.
> 
> In larger environments with many OSDs per node, the cluster has enough redundancy to
> tolerate multiple sequential OSD restarts on the same node without risk, so using the
> loop is safe.
> 
> :::

Reviewer: The "small environment / many OSDs per node" framing is misleading — the actual safety criterion is node count and CRUSH failure domain, not OSDs per node. With host-level failure domain (the OSISM default) and ≥ 3 nodes, restarting all OSDs on one node at once is always safe: the other two nodes retain enough replicas to satisfy `min_size`. Also, the inter-node criterion should be "all PGs `active+clean`" rather than `HEALTH_OK`, since `HEALTH_WARN` for stray daemons is expected throughout migration.

Note: if this suggestion is accepted, the earlier warning block added by the companion suggestion (moving it before the adopt commands) should be removed.

```suggestion
:::warning

Adopt OSDs **one node at a time**. After completing all OSDs on a node, wait until all
PGs return to `active+clean` (`ceph -s` shows all PGs `active+clean` under `data:`)
before proceeding to the next node. The overall health may still show `HEALTH_WARN` for
stray daemons during migration — this is expected and not a stop condition.

Within a single node, it is safe to use the loop above without waiting between individual
OSDs, provided your cluster meets both of the following conditions:

* At least **3 OSD nodes**.
* **Host-level CRUSH failure domain** — the OSISM default. Verify with:
  ```bash
  ceph osd crush rule dump
  ```
  All rules should show `"type": "host"` in the `chooseleaf` step.

If your cluster has only 2 OSD nodes, or any rule shows a different failure domain type,
adopt each OSD individually and wait for all PGs to return to `active+clean` after each
one before continuing.

:::
```

NAME   HOST            PORTS  STATUS          REFRESHED  AGE  MEM USE  MEM LIM  VERSION  IMAGE ID      CONTAINER ID
osd.0  testbed-node-3         running (46s)      0s ago    -     181M    4096M  18.2.8   01985efead8e  f83bb9204db5
osd.1  testbed-node-4         running (5m)       0s ago   2m     167M    4096M  18.2.8   01985efead8e  11cd8d77a78b
osd.2  testbed-node-5         running (118s)     0s ago    -     176M    4096M  18.2.8   01985efead8e  f608633171a8
osd.3  testbed-node-4         running (3m)       0s ago   2m     196M    4096M  18.2.8   01985efead8e  5032744c6063
osd.4  testbed-node-5         running (2m)       0s ago    -     179M    4096M  18.2.8   01985efead8e  a9c9f18801d9
osd.5  testbed-node-3         running (38s)      0s ago    -     153M    4096M  18.2.8   01985efead8e  bfd02ac996db
```

Once all PGs are `active+clean`, remove the safety flags and re-enable the PG
autoscaler on the OSISM manager node:

```bash
ceph osd unset noout
ceph osd unset nodeep-scrub
ceph balancer on
```

```bash
if [ -f /home/dragon/autoscale_pools.txt ]; then
    while read pool; do
        ceph osd pool set ${pool} pg_autoscale_mode on
    done < /home/dragon/autoscale_pools.txt

>     rm /home/dragon/autoscale_pools.txt

Reviewer: If something goes wrong mid-loop, `autoscale_pools.txt` is deleted before all pools are re-enabled, leaving no record for a retry. The `rm` should only happen once the user has verified all pools succeeded:

```suggestion
if [ -f /home/dragon/autoscale_pools.txt ]; then
    while read pool; do
        if ceph osd pool set ${pool} pg_autoscale_mode on > /dev/null 2>&1; then
            echo "OK: ${pool}"
        else
            echo "FAILED: ${pool}"
        fi
    done < /home/dragon/autoscale_pools.txt
fi
# Once all pools show OK:, run: rm /home/dragon/autoscale_pools.txt
```

ceph orch apply crash
```

Verify that the crash daemons are running:

```bash
ceph orch ls
```

The output should now include the crash service:

```console
NAME   PORTS  RUNNING  REFRESHED  AGE  PLACEMENT
crash             6/6  3s ago     23s  *
mgr               3/0  3s ago     -    <unmanaged>
mon               3/0  3s ago     -    <unmanaged>
osd                 6  3s ago     -    <unmanaged>
```

### Migrating RGW daemons

:::note

The migration of RGW has currently only been tested for single-site deployments.
Instructions for multi-site RGW setups will be added in a future update of this guide.

:::

RGW daemons cannot be adopted in-place with `cephadm adopt`. Instead, new RGW daemons
are deployed via the orchestrator and the legacy daemons are stopped afterwards.

Determine the RGW nodes on the OSISM manager node:

```bash
osism get hosts -l ceph-rgw
```

Determine the RGW realm, zone group, and zone name from the running cluster:

```bash
radosgw-admin realm list
radosgw-admin zonegroup list
radosgw-admin zone list
```

In a typical single-site deployment, the default values are `default` for all three.
The service ID for the `ceph orch apply rgw` command is composed as

> `<realm_name>.<zone_name>` (e.g. `default.default`).

Reviewer: In a standard OSISM testbed the realm is empty (`"realms": []`), so the description "default for all three" is misleading. The script already handles this with a `|| echo "default"` fallback, but the description should reflect reality:

```suggestion
In a typical single-site OSISM deployment, the zone and zonegroup are `default`. The realm
may be empty — the script below falls back to `default` in that case, giving a service ID
of `default.default`.
The service ID for the `ceph orch apply rgw` command is composed as
`<realm_name>.<zone_name>` (e.g. `default.default`).
```

# Deploy new MDS daemons via the orchestrator
ceph orch apply mds ${CEPHFS_NAME} --placement="${MDS_PLACEMENT}"

# Wait until the new MDS daemons are running
echo "Waiting for new MDS daemons..."
until ceph orch ps --daemon-type mds --format json | python3 -c "
import json, sys
daemons = json.load(sys.stdin)
running = [d for d in daemons if d.get('status_desc') == 'running']
sys.exit(0 if len(running) >= len('${MDS_PLACEMENT}'.split(',')) else 1)
" 2>/dev/null; do
    sleep 5
done
echo "New MDS daemons are running."
```

Verify that the new MDS daemons are running:

```bash
ceph orch ls --service-type mds
```

The output should look similar to this:

```console
NAME        PORTS  RUNNING  REFRESHED  AGE  PLACEMENT
mds.cephfs             3/3  0s ago     13s  testbed-node-3;testbed-node-4;testbed-node-5
```

Then stop and disable the legacy MDS daemons. Run the following on the OSISM manager node:

```bash
for node in $(osism get hosts -l ceph-mds | awk 'NR>3 && /\|/ {print $2}'); do
    echo "Stopping legacy MDS daemon on ${node}..."
    ssh ${node} "sudo systemctl stop ceph-mds@${node}.service; sudo systemctl disable ceph-mds@${node}.service"
done
```

> ## Step 6: Apply service specs

Reviewer: As far as I can tell, Step 6 is optional and currently broken. My preferred solution is working automation, but in the interest of getting this guide merged soon, I recommend removing Step 6 entirely. We can always add it back once we have a working solution.

After adoption, daemons show as `<unmanaged>` in `ceph orch ls` because the orchestrator
has no service spec for them. Apply service specs on the OSISM manager node to make them
managed:

```bash
MON_PLACEMENT=$(osism get hosts -l ceph-mon | awk 'NR>3 && /\|/ {print $2}' | paste -sd,)
MGR_PLACEMENT=$(osism get hosts -l ceph-mgr | awk 'NR>3 && /\|/ {print $2}' | paste -sd,)

ceph orch apply mon --placement="${MON_PLACEMENT}"
ceph orch apply mgr --placement="${MGR_PLACEMENT}"
```

:::note

Adopted OSD daemons will show as `<unmanaged>` in `ceph orch ls`. This is expected
and does not affect their operation — the OSDs are fully functional, they are just not
governed by an orchestrator service spec.

To transition the adopted OSDs to a managed state, create an OSD service spec with
device filters that match exactly the devices already in use. For example, if all OSDs
use the same device path pattern:

```yaml
service_type: osd
service_id: default_drive_group
placement:
  host_pattern: '*'
data_devices:
  paths:
    - /dev/sdb

>     - /dev/sdc

Reviewer: Yes, such a file can be built from `ceph osd tree` and `ceph device ls`, but it may not be obvious to users who have not done it before. I wonder if we can offer them some help (more information or a script).

```

Apply the spec:

```bash

> ceph orch apply -i osd_spec.yaml

Reviewer: This would need the same pattern used for the ssh keys to mount the file into the container, otherwise the file cannot be found. Something like:

```
cp osd_spec.yaml /opt/cephclient/data/
ceph orch apply -i /data/osd_spec.yaml
rm /opt/cephclient/data/osd_spec.yaml
```

But even with this, "ceph orch ls" showed my osds as "unmanaged" (plus a managed osd.default_drive_group with 0 members). Maybe we should advise users to leave the osds unmanaged for now.

```json
{
    "mon": {
        "ceph version 18.2.8 (efac5a54607c13fa50d4822e50242b86e6e446df) reef (stable)": 3
    },
    "mgr": {
        "ceph version 18.2.8 (efac5a54607c13fa50d4822e50242b86e6e446df) reef (stable)": 3
    },
    "osd": {
        "ceph version 18.2.8 (efac5a54607c13fa50d4822e50242b86e6e446df) reef (stable)": 6
    },
    "mds": {
        "ceph version 18.2.8 (efac5a54607c13fa50d4822e50242b86e6e446df) reef (stable)": 3
    },
    "rgw": {
        "ceph version 18.2.8 (efac5a54607c13fa50d4822e50242b86e6e446df) reef (stable)": 3
    },
    "overall": {
        "ceph version 18.2.8 (efac5a54607c13fa50d4822e50242b86e6e446df) reef (stable)": 18
    }
}
```

## Step 8: Clean up ceph-ansible artifacts

Only proceed with this step after completing the verification in
[Step 7](#step-7-verify-the-migration): `ceph -s` shows `HEALTH_OK`, all daemons
appear as `running` in `ceph orch ps`, and `ceph versions` reports a single consistent
version across all daemon types.

1. Remove old systemd unit files that are no longer used. Do **not** use a wildcard
   like `ceph-*.service` as this would also remove the cephadm-managed units

>    (e.g. `ceph-<FSID>@.service`). Remove only the legacy units:

Reviewer: ```suggestion
1. Remove old systemd unit files that are no longer used. Run on **each Ceph node**.
   Do **not** use a wildcard like `ceph-*.service` as this would also remove the
   cephadm-managed units (e.g. `ceph-<FSID>@.service`). Remove only the legacy units:
```

              /etc/systemd/system/ceph-mds@.service \
              /etc/systemd/system/ceph-radosgw@.service \
              /etc/systemd/system/ceph-crash@.service \
              /etc/systemd/system/ceph-mon.target \
              /etc/systemd/system/ceph-mgr.target \
              /etc/systemd/system/ceph-osd.target \
              /etc/systemd/system/ceph-mds.target \
              /etc/systemd/system/ceph-radosgw.target \
              /etc/systemd/system/ceph.target
   sudo rm -rf /etc/systemd/system/ceph.target.wants
   sudo systemctl daemon-reload
   ```

2. The ceph-ansible configuration in the OSISM configuration repository
   (`environments/ceph/`) can be kept as a reference but is no longer active.

:::note

Do **not** remove `/etc/ceph/ceph.conf` or any keyring files. These are still
used by the cluster and are now managed by cephadm.

:::

## Post-migration operations

After migration, Ceph lifecycle operations are performed through cephadm and the
Ceph orchestrator instead of `osism apply ceph-*` commands.

| ceph-ansible (before)             | cephadm (after)                         |
|:----------------------------------|:----------------------------------------|

> | `osism apply ceph-mons`           | `ceph orch apply mon`                   |

Reviewer: What happens if a user issues an old `osism apply ceph-*` command out of habit? Ideally, running `osism apply ceph-mons` should now tell the user something like "Please run `ceph orch apply mon`", but the command looks like it is running as before. Maybe we should tell users if this is is dangerous or not and if they can safely abort the commands once they realize their mistake. -- My preferred solution would be to disable the old commands during the upgrade, but as long as they remain present, we should document what happens if they are used.

| Squid (19)   | `cephadm_19.2.3-0ubuntu0.24.04.2~cloud0_amd64.deb` |

These are the only versions available in the UCA. The exact point release does not need
to match your running Ceph version -- only the release series (Quincy, Reef, Squid) matters.

```bash
CEPHADM_PKG=cephadm_18.2.4-0ubuntu1~cloud1_amd64.deb
```

```bash
curl --silent --remote-name --location https://ubuntu-cloud.archive.canonical.com/ubuntu/pool/main/c/ceph/${CEPHADM_PKG}
sudo dpkg -i ${CEPHADM_PKG}
```

### OSD nodes

The UCA cephadm package for Reef (18.2.4) contains a
[known bug](https://tracker.ceph.com/issues/63151) that causes a crash when parsing
AppArmor profiles during OSD adoption. On OSD nodes, install cephadm as a standalone
Python script from the Ceph Git repository instead:

```bash
CEPH_RELEASE=$(docker inspect $(docker ps --filter "name=ceph" --format "{{.Names}}" | head -1) --format '{{.Config.Image}}' | cut -d: -f2)
```

```bash
curl --silent --remote-name --location https://raw.githubusercontent.com/ceph/ceph/${CEPH_RELEASE}/src/cephadm/cephadm.py
chmod +x cephadm.py
sudo mv cephadm.py /usr/sbin/cephadm
```

### Verify

On each node, verify the installation:

```bash
sudo cephadm version

> ```

Reviewer 2: Why is this needed?

## Step 3: Prepare the cephadm configuration

On each Ceph node, prepare the host for cephadm:

```bash
sudo cephadm prepare-host
```

Confirm that the existing `/etc/ceph/ceph.conf` and `/etc/ceph/ceph.client.admin.keyring`
are accessible on the Ceph nodes. Cephadm will use these to connect to the cluster.

On each Ceph node, determine the currently used container image from a running Ceph
daemon:

```bash
CEPH_IMAGE=$(docker inspect $(docker ps --filter "name=ceph" --format "{{.Names}}" | head -1) --format '{{.Config.Image}}')
```

Set the container image in the Ceph configuration. Run the following command on the
manager node:

```bash
ceph config set global container_image ${CEPH_IMAGE}
```

> ## Step 4: Enable the cephadm orchestrator

Reviewer 2: ```suggestion
## Step 4: Configure cephadm
```

```

Register all Ceph nodes with the orchestrator on the manager node:

```bash
for node in $(osism get hosts -l ceph | awk 'NR>3 && /\|/ {print $2}'); do
    ceph orch host add ${node} $(getent hosts ${node} | awk '{print $1}')
done
```

Verify that all hosts have been registered:

```bash
ceph orch host ls
```

## Step 5: Adopt daemons

Adopt all Ceph daemons across the cluster. The recommended order is:

1. **Monitors** (mon)
2. **Managers** (mgr)
3. **OSDs** (osd)

For each daemon, run the adopt command on the respective node. Set the variables
on each node first:

```bash
CEPH_HOSTNAME=$(hostname)
CEPH_IMAGE=$(docker inspect $(docker ps --filter "name=ceph" --format "{{.Names}}" | head -1) --format '{{.Config.Image}}')
```

:::warning

Verify cluster health with `ceph -s` on the manager node after each step.
Do not proceed if the cluster is not healthy.

:::

### Adopting monitor daemons

```bash
sudo cephadm --image ${CEPH_IMAGE} adopt --style legacy --skip-firewalld --name mon.${CEPH_HOSTNAME}
```

### Adopting manager daemons

```bash
sudo cephadm --image ${CEPH_IMAGE} adopt --style legacy --skip-firewalld --name mgr.${CEPH_HOSTNAME}
```

### Verify monitors and managers

> After all monitors and managers have been adopted, verify on the manager node that

Reviewer 2: We should clarify which "manager node" is meant here.

   sudo rm -f /etc/systemd/system/ceph-*.service
   sudo systemctl daemon-reload
   ```

2. The ceph-ansible configuration in the OSISM configuration repository
   (`environments/ceph/`) can be kept as a reference but is no longer active.

:::note

Do **not** remove `/etc/ceph/ceph.conf` or any keyring files. These are still
used by the cluster and are now managed by cephadm.

:::

## Post-migration operations

After migration, Ceph lifecycle operations are performed through cephadm and the
Ceph orchestrator instead of `osism apply ceph-*` commands.

| ceph-ansible (before)             | cephadm (after)                         |
|:----------------------------------|:----------------------------------------|
| `osism apply ceph-mons`           | `ceph orch apply mon`                   |
| `osism apply ceph-mgrs`           | `ceph orch apply mgr`                   |
| `osism apply ceph-osds`           | `ceph orch apply osd`                   |
| `osism apply ceph-rgws`           | `ceph orch apply rgw`                   |
| `osism apply ceph-mdss`           | `ceph orch apply mds`                   |
| Editing `configuration.yml`       | `ceph config set <section> <key> <val>` |

> | `osism apply ceph-rolling_update` | `ceph orch upgrade start --image <img>` |

Reviewer 2: What does that mean? Are they or are they not 1:1 mappings?

> | `osism apply ceph-rolling_update` | `ceph orch upgrade start --image <img>` |

Author: 1:1 mappings.

:::note

The migration of RGW has currently only been tested for single-site deployments.
Instructions for multi-site RGW setups will be added in a future update of this guide.

:::

RGW daemons cannot be adopted in-place with `cephadm adopt`. Instead, new RGW daemons
are deployed via the orchestrator and the legacy daemons are stopped afterwards.

Determine the RGW nodes on the manager node:

```bash
osism get hosts -l ceph-rgw
```

Determine the RGW frontend port from the ceph-ansible configuration
(`environments/ceph/configuration.yml`). The variable `radosgw_frontend_port` contains
the port (default: `8081`).

Deploy new RGW daemons on the manager node using a service spec. Replace the
placeholders with the actual values:

```bash
ceph orch apply rgw <service_id> --placement="<rgw-node-1>,<rgw-node-2>,<rgw-node-3>" --port=<radosgw_frontend_port>
```

If the RGW service was configured with SSL (i.e. `radosgw_frontend_ssl_certificate` is
set in the ceph-ansible configuration), the SSL certificate must be imported into the
Ceph config-key store before deploying. Run this on each RGW node:

```bash
sudo ceph config-key set rgw/cert/rgw.$(hostname) -i <path_to_ssl_certificate>
```

> Then add `--ssl` to the `ceph orch apply rgw` command.

Reviewer 2: maybe this should be mentioned before we run the `ceph orch apply rgw...` command above? people tend to run things first and then continue reading what is said below
